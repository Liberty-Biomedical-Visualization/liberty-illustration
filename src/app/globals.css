@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
