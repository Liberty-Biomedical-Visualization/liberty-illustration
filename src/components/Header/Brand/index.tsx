import Image from "next/image";

import type { ImageData } from "@/lib/content";

export default function Brand(props: BrandProps) {
  const { className, logoData } = props;
  const { description, height, src, title, width } = logoData;

  return (
    <div className={className}>
      <Image
        alt={description ?? ""}
        className="max-w-3/4 mx-auto"
        height={height}
        src={src}
        title={title}
        priority
        width={width}
      />
      <h1 className="font-light sm:text-3xl text-2xl text-center text-nowrap">
        Abigail Richbourg Liberty, MS, CMI
      </h1>
    </div>
  );
}

export interface BrandProps {
  className?: string;
  logoData: ImageData;
}
