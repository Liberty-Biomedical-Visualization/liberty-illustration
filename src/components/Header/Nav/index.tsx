"use client";

import FocusTrap from "focus-trap-react";
import { usePathname } from "next/navigation";
import { useCallback, useEffect, useRef, useState } from "react";

import resolveClassNames from "@/lib/resolveClassNames";

import NavList, { type NavListProps } from "./List";
import NavListToggle from "./ListToggle";

export default function Nav(props: Readonly<NavProps>) {
  const { disableTabbableDisplayCheck, pages } = props;
  const [navIsShown, setNavIsShown] = useState(false);
  const pathname = usePathname();
  const lastPathname = useRef(pathname);
  const bodyPaddingRight = useRef(0);

  const hideNav = useCallback(() => {
    enableScrolling(bodyPaddingRight.current);
    setNavIsShown(false);
  }, []);

  const hideNavOnNavigation = useCallback(() => {
    if (navIsShown && lastPathname.current !== pathname) {
      hideNav();
    }
  }, [hideNav, lastPathname, navIsShown, pathname]);

  useEffect(hideNavOnNavigation, [hideNavOnNavigation]);

  const updateLastPathname = () => {
    lastPathname.current = pathname;
  };

  useEffect(updateLastPathname, [pathname]);

  useEffect(
    () => hideNavWhenShownAboveMediumBreakpoint(hideNav, navIsShown),
    [hideNav, navIsShown],
  );

  const showNav = () => {
    bodyPaddingRight.current = parseFloat(
      getComputedStyle(document.body).paddingRight,
    );
    disableScrolling(bodyPaddingRight.current);
    setNavIsShown(true);
  };

  const toggleList = () => {
    if (navIsShown) {
      hideNav();
    } else {
      showNav();
    }
  };

  const handleEscape = () => {
    hideNav();
    return true;
  };

  const hideNavOnSamePage = (href: string) => {
    if (href === pathname) {
      hideNav();
    }
  };

  const className = resolveClassNames(
    baseClassNames,
    navIsShown ? "visible" : "hidden",
    props.className,
  );

  return (
    <FocusTrap
      active={navIsShown}
      focusTrapOptions={{
        escapeDeactivates: handleEscape,
        tabbableOptions: {
          // Setting displayCheck to "none" allows FocusTrap to be used with
          // jsdom for testing.
          displayCheck: disableTabbableDisplayCheck ? "none" : "full",
        },
      }}
    >
      <div>
        <NavListToggle
          id={navListToggleId}
          listIsShown={navIsShown}
          navListId={navId}
          onClick={toggleList}
        />
        <nav aria-labelledby={navListToggleId} className={className} id={navId}>
          <NavList onItemClick={hideNavOnSamePage} pages={pages} />
        </nav>
      </div>
    </FocusTrap>
  );
}

export interface NavProps {
  className?: string;
  disableTabbableDisplayCheck?: boolean | undefined;
  pages: NavListProps["pages"];
}

const baseClassNames =
  "bg-white fixed h-full left-0 md:flex md:h-auto md:static md:visible top-0 w-full z-10";

const navId = "nav";
const navListToggleId = "nav-list-toggle";

function disableScrolling(bodyPaddingRight: number) {
  const scrollbarWidth =
    window.innerWidth - document.documentElement.clientWidth;
  document.body.style.overflow = "hidden";
  document.body.style.paddingRight = bodyPaddingRight + scrollbarWidth + "px";
}

function enableScrolling(bodyPaddingRight: number) {
  document.body.style.overflow = "auto";
  document.body.style.paddingRight = bodyPaddingRight + "px";
}

function hideNavWhenShownAboveMediumBreakpoint(
  hideList: () => void,
  listIsShown: boolean,
) {
  // Guard against running an environment without a global window such as tests.
  if (
    typeof window === "undefined" ||
    typeof window.matchMedia !== "function"
  ) {
    return;
  }

  const mediaQuery = window.matchMedia("(max-width: 768px)");

  const handleMediaQueryChange = () => {
    if (listIsShown && !mediaQuery.matches) {
      hideList();
    }
  };

  mediaQuery.addEventListener("change", handleMediaQueryChange);

  const stopHidingListWhenShownAboveMediumBreakpoint = () =>
    mediaQuery.removeEventListener("change", handleMediaQueryChange);

  return stopHidingListWhenShownAboveMediumBreakpoint;
}
