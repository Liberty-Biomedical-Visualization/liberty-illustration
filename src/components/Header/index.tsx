import resolveClassNames from "@/lib/resolveClassNames";

import Brand, { type BrandProps } from "./Brand";

export default function Header(props: HeaderProps) {
  const { logoData } = props;
  const className = resolveClassNames(
    "flex flex-col items-center",
    props.className,
  );

  return (
    <header className={className}>
      <Brand logoData={logoData} />
    </header>
  );
}

export interface HeaderProps {
  className?: string;
  logoData: BrandProps["logoData"];
}
