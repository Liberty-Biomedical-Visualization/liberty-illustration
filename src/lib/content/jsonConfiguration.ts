/**
 * Expected structure of JSON entries in Contentful.
 */
export const jsonConfiguration = {
  "Home Configuration": {
    carouselImageGalleryId: "string",
    slideDuration: "number",
  },
  "Integration Test JSON": {
    foo: "string",
  },
  "Invalid Integration Test JSON": {
    fizz: "string[]",
    foo: "string",
    zip: "string[]",
  },
  "Medical Illustration Configuration": {
    imageGalleryId: "string",
  },
  "Portfolio Configuration": {
    imageGalleryIds: "string[]",
  },
  "Site Configuration": {
    amiLogoId: "string",
    cmiLogoId: "string",
    copyrightYear: "number",
    navPageOrder: "string[]",
    siteLogoId: "string",
  },
  "Site Metadata": {
    author: "string",
    description: "string",
    faviconId: "string",
    keywords: "string[]",
    locale: "string",
    name: "string",
    url: "string",
  },
} as const;

export type JsonTitle = keyof JsonConfiguration;
type JsonConfiguration = typeof jsonConfiguration;

export type Json<K extends JsonTitle> = StringLiteralPropertiesAsTypes<
  JsonConfiguration[K]
>;

type StringLiteralPropertiesAsTypes<T extends JsonConfiguration[JsonTitle]> = {
  [K in keyof T]: StringLiteralAsType<T[K]>;
};

type StringLiteralAsType<T> = T extends ConfigurableType
  ? ConfigurableTypeMap[T]
  : never;

type ConfigurableType = keyof ConfigurableTypeMap;

type ConfigurableTypeMap = {
  number: number;
  string: string;
  "string[]": string[];
};
