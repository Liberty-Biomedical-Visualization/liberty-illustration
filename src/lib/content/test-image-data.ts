import config from "../config";
import type { ImageData } from ".";

export const completeImage: Readonly<Required<ImageData>> = {
  description: "A complete image used for testing.",
  height: 100,
  src: "https://foo.com/complete.jpg",
  title: "Complete Image",
  width: 100,
};

export const describedImage: Readonly<ImageData & { description: string }> = {
  description: "A described image used for testing.",
  height: 100,
  src: "https://foo.com/described.jpg",
  width: 100,
};

export const imageA: Readonly<Required<ImageData>> = {
  description: "An image identified as A and used for testing.",
  height: 100,
  src: "https://foo.com/a.jpg",
  title: "Image A",
  width: 100,
};

export const imageB: Readonly<Required<ImageData>> = {
  description: "An image identified as B and used for testing.",
  height: 100,
  src: "https://foo.com/b.jpg",
  title: "Image B",
  width: 100,
};

export const imageC: Readonly<Required<ImageData>> = {
  description: "An image identified as C and used for testing.",
  height: 100,
  src: "https://foo.com/c.jpg",
  title: "Image C",
  width: 100,
};

export const integrationTestImage: Readonly<Required<ImageData>> = {
  description: "This is a dummy image used for integration testing.",
  height: 2_480,
  src: `https://${config.CONTENTFUL_IMAGE_CDN_HOSTNAME}/${config.CONTENTFUL_SPACE_ID}/7ci4trPV2FWGxweCTSVqol/b6e816c40b1cf429be48c82a0f73037c/test-image.png`,
  title: "Integration Test Image",
  width: 2_480,
};

export const longDescribedImage: Readonly<ImageData & { description: string }> =
  {
    description:
      "This Is An Image With An Exceedingly Long Description That Will Be Truncated When Converted to a Thumbnail Because It Is Longer Than the Maximum Character Count",
    height: 100,
    src: "https://foo.com/long-titled.jpg",
    width: 100,
  };

export const longTitledImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/long-titled.jpg",
  title:
    "An Image With An Exceedingly Long Title That Will Be Truncated When Converted to a Thumbnail Because It Is Longer Than the Maximum Character Count",
  width: 100,
};

export const minimalImage: Readonly<ImageData> = {
  height: 100,
  src: "https://foo.com/minimal.jpg",
  width: 100,
};

export const nondescriptImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/nondescript.jpg",
  title: "Nondescript Image",
  width: 100,
};

export const titledImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/titled.jpg",
  title: "Titled Image",
  width: 100,
};

export const untitledImage: Readonly<ImageData & { description: string }> = {
  description: "An untitled image used for testing.",
  height: 100,
  src: "https://foo.com/untitled.jpg",
  width: 100,
};
