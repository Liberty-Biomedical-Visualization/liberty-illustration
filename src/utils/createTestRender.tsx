import { render } from "@testing-library/react";
import type { JSX, JSXElementConstructor } from "react";

/**
 * Returns a function that allows rendering Component with partial props while
 * falling back to defaultProps for testing purposes.
 */
export default function createTestRender<C extends JSXElementConstructor<P>, P>(
  Component: JSXElementConstructor<P> & C,
  defaultProps: JSX.LibraryManagedAttributes<C, P>,
) {
  return function (props: Partial<P> = {}) {
    const resolvedProps: JSX.IntrinsicAttributes &
      JSX.LibraryManagedAttributes<C, P> = { ...defaultProps, ...props };
    return render(<Component {...resolvedProps} />);
  };
}
