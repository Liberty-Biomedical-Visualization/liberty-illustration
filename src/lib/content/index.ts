import * as contentful from "contentful";

import packageData from "../../../package.json";
import config from "../config";
import isDefined from "../isDefined";
import { Json, JsonTitle } from "./jsonConfiguration";
import { validateJson } from "./validateJson";

/**
 * Gets `AssetData` by ID.
 *
 * Throws an `Error` if the asset is not found.
 */
export async function getAssetData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToAssetData(asset);
}

export interface AssetData {
  description?: string;
  title?: string;
  url: string;
}

/**
 * Gets `ImageData` by ID.
 *
 * Throws an `Error` if the image is not found, or the asset is not an image.
 */
export async function getImageData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToImageData(asset);
}

export interface ImageData {
  description?: string;
  height: number;
  src: string;
  title?: string;
  width: number;
}

/**
 * Gets an `ImageGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getImageGallery(id: string) {
  const entry = await client.getEntry<ImageGallerySkeleton>(id);
  return mapEntryToImageGallery(entry);
}

export interface ImageGallery {
  images: ImageData[];
  title: string;
}

/**
 * Gets a `JsonReference` by title.
 *
 * Throws an error if the JSON is not found or not properly structured.
 */
export async function getJsonByTitle<K extends JsonTitle>(title: K) {
  const jsons = await client.getEntries<JsonSkeleton<Json<K>>>({
    content_type: "json",
    "fields.title": title,
  });
  const json = jsons.items[0];

  if (json === undefined) {
    throw new Error(`JSON with title "${title}" not found`);
  }

  validateJson(json.fields.value, title);
  return json.fields.value;
}

/**
 * Gets a `VideoGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getVideoGallery(id: string) {
  const entry = await client.getEntry<VideoGallerySkeleton>(id);
  return mapEntryToVideoGallery(entry);
}

export interface VideoGallery {
  title: string;
  videos: VideoData[];
}

export interface VideoData {
  title: string;
  url: string;
}

function mapContentfulAssetToAssetData(asset: ContentfulAsset) {
  const { description, file, title } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const assetData: AssetData = { url: "https:" + file.url };

  if (description) {
    assetData.description = description;
  }

  if (title) {
    assetData.title = title;
  }

  return assetData;
}

function mapContentfulAssetToImageData(asset: ContentfulAsset) {
  const { file } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const { image } = file.details;

  if (image === undefined) {
    throw new Error("image is undefined");
  }

  const { height, width } = image;
  const { description, title, url } = mapContentfulAssetToAssetData(asset);
  const imageData: ImageData = { height, src: url, width };

  if (description) {
    imageData.description = description;
  }

  if (title) {
    imageData.title = title;
  }

  return imageData;
}

type ContentfulAsset = contentful.Asset<"WITHOUT_UNRESOLVABLE_LINKS", string>;

function mapEntryToImageGallery(entry: ImageGalleryEntry) {
  const { images, title } = entry.fields;
  const mappedImages = images
    .filter(isDefined)
    .map(mapContentfulAssetToImageData);
  const gallery: ImageGallery = { images: mappedImages, title };
  return gallery;
}

type ImageGalleryEntry = contentful.Entry<
  ImageGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface ImageGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
    title: contentful.EntryFieldTypes.Text;
  };
}

function mapEntryToVideoGallery(entry: VideoGalleryEntry) {
  const { title, videos } = entry.fields;
  const mappedVideos = videos.filter(isDefined).map((video) => video.fields);
  const gallery: VideoGallery = { title, videos: mappedVideos };
  return gallery;
}

type VideoGalleryEntry = contentful.Entry<
  VideoGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface VideoGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    videos: contentful.EntryFieldTypes.Array<
      contentful.EntryFieldTypes.EntryLink<VideoLinkSkeleton>
    >;
  };
}

interface VideoLinkSkeleton {
  contentTypeId: "videoLink";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    url: contentful.EntryFieldTypes.Text;
  };
}

interface JsonSkeleton<T extends JsonReference> {
  contentTypeId: "json";
  fields: {
    value: contentful.EntryFieldTypes.Object<T>;
    title: contentful.EntryFieldTypes.Text;
  };
}

type JsonReference = JsonArray | JsonObject;
type JsonArray = JsonValue[];
type JsonObject = { [key: string]: JsonValue };
type JsonValue = JsonArray | JsonObject | boolean | null | number | string;

const { dependencies, name, version } = packageData;

const client = contentful.createClient({
  accessToken: config.CONTENTFUL_ACCESS_TOKEN,
  application: `${name}/${version}`,
  environment: config.CONTENTFUL_ENVIRONMENT,
  integration: `next/${dependencies.next}`,
  space: config.CONTENTFUL_SPACE_ID,
}).withoutUnresolvableLinks;
