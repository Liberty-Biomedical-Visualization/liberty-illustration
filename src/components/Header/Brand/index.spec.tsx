import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";

import { describedImage, minimalImage } from "@/lib/content/test-image-data";
import createTestRender from "@/lib/createTestRender";

import Brand from ".";

describe("Brand", () => {
  it("should display the top-level heading", () => {
    renderBrand();
    const result = screen.getByRole("heading", {
      level: 1,
      name: "Abigail Richbourg Liberty, MS, CMI Biomedical illustration • Animation • Graphic design",
    });
    expect(result).toBeVisible();
  });

  it("should display an accessible logo when passed image data with a description", () => {
    renderBrand({ logoData: describedImage });
    const result = screen.getByRole("img", {
      name: describedImage.description,
    });
    expect(result).toBeVisible();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderBrand({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderBrand();
    expect(result.container).toMatchSnapshot();
  });
});

const renderBrand = createTestRender(Brand, {
  logoData: minimalImage,
});
