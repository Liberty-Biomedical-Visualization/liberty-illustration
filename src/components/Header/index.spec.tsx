import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";

import { describedImage, minimalImage } from "@/lib/content/test-image-data";
import createTestRender from "@/lib/createTestRender";

import Header from ".";

describe("Header", () => {
  it("should display the site banner", () => {
    renderHeader();
    const result = screen.getByRole("banner");
    expect(result).toBeVisible();
  });

  it("should display an accessible logo when passed image data with a description", () => {
    renderHeader({ logoData: describedImage });
    const result = screen.getByRole("img", {
      name: describedImage.description,
    });
    expect(result).toBeVisible();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderHeader({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderHeader();
    expect(result.container).toMatchSnapshot();
  });
});

const renderHeader = createTestRender(Header, {
  logoData: minimalImage,
});
