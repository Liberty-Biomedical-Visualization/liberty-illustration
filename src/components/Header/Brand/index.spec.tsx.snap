// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brand should match the snapshot 1`] = `
<div>
  <div>
    <img
      alt=""
      class="max-w-3/4 mx-auto"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="100"
      src="/_next/image?url=https%3A%2F%2Ffoo.com%2Fminimal.jpg&w=256&q=75"
      srcset="/_next/image?url=https%3A%2F%2Ffoo.com%2Fminimal.jpg&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Ffoo.com%2Fminimal.jpg&w=256&q=75 2x"
      style="color: transparent;"
      width="100"
    />
    <h1
      class="font-light sm:text-3xl text-2xl text-center text-nowrap"
    >
      Abigail Richbourg Liberty, MS, CMI 
      <br />
      <span
        class="sm:text-lg text-sm"
      >
        Biomedical illustration • Animation • Graphic design
      </span>
    </h1>
  </div>
</div>
`;
